File: src/stores/sector.ts
import { defineStore } from 'pinia';
import { computed, ref } from 'vue';
import { invoke } from '@tauri-apps/api/core';
import type { RotationStatus, SectorKind, SectorLimitUpStats, SectorMember, SectorMemberPage, SectorRotation, SectorSummary, SectorSummaryPage } from '@/types/sector';

export const useSectorStore = defineStore('sector', () => {
  const kind = ref<SectorKind>('industry');
  const keyword = ref('');
  const page = ref(1);
  const pageSize = ref(50);
  const rows = ref<SectorSummary[]>([]);
  const total = ref(0);
  const asOf = ref('');
  const source = ref('');
  const stale = ref(false);
  const loading = ref(false);
  const error = ref<string | null>(null);
  const selected = ref<SectorSummary | null>(null);
  const members = ref<SectorMember[]>([]);
  const memberPage = ref(1);
  const memberPageSize = ref(50);
  const memberTotal = ref(0);
  const memberAsOf = ref('');
  const memberSource = ref('');
  const memberStale = ref(false);
  const memberLoading = ref(false);
  const memberError = ref<string | null>(null);
  const limitUpStats = ref<SectorLimitUpStats | null>(null);
  const limitUpError = ref<string | null>(null);
  const rotationRows = ref<SectorSummary[]>([]);
  const rotationStatuses = ref<RotationStatus[]>([]);
  const rotationSource = ref('');
  const rotationLoading = ref(false);
  const rotationError = ref<string | null>(null);
  let generation = 0;
  let memberGeneration = 0;

  const pageCount = computed(() => Math.max(1, Math.ceil(total.value / pageSize.value)));
  const memberPageCount = computed(() => Math.max(1, Math.ceil(memberTotal.value / memberPageSize.value)));

  async function fetchSummaries(forceRefresh = false) {
    const request = ++generation;
    loading.value = true;
    error.value = null;
    try {
      const response = await invoke<SectorSummaryPage>('get_sector_summaries', {
        kind: kind.value,
        page: page.value,
        pageSize: pageSize.value,
        keyword: keyword.value,
        forceRefresh,
      });
      if (request !== generation) return;
      rows.value = response.items;
      total.value = response.total;
      page.value = response.page;
      pageSize.value = response.page_size;
      asOf.value = response.as_of;
      source.value = response.source;
      stale.value = response.stale;
    } catch (e) {
      if (request === generation) error.value = `获取板块排行失败: ${e}`;
      console.error('[sector] fetchSummaries failed:', e);
    } finally {
      if (request === generation) loading.value = false;
    }
  }

  async function selectKind(next: SectorKind) {
    if (kind.value === next) return;
    kind.value = next;
    page.value = 1;
    await fetchSummaries();
  }

  async function selectPage(next: number) {
    page.value = Math.max(1, next);
    await fetchSummaries();
  }

  async function selectSector(row: SectorSummary) {
    selected.value = row;
    memberPage.value = 1;
    members.value = [];
    memberError.value = null;
    limitUpStats.value = null;
    limitUpError.value = null;
    await Promise.all([fetchMembers(), fetchLimitUpStats()]);
  }

  function clearSector() {
    memberGeneration += 1;
    selected.value = null;
    members.value = [];
    memberError.value = null;
    limitUpStats.value = null;
    limitUpError.value = null;
    memberLoading.value = false;
  }

  async function fetchLimitUpStats() {
    const target = selected.value;
    if (!target) return;
    try {
      const response = await invoke<SectorLimitUpStats>('get_sector_limit_up_stats', { sectorCode: target.code });
      if (selected.value?.code === target.code) limitUpStats.value = response;
    } catch (e) {
      if (selected.value?.code === target.code) limitUpError.value = String(e);
    }
  }

  async function fetchMembers(forceRefresh = false) {
    const target = selected.value;
    if (!target) return;
    const request = ++memberGeneration;
    memberLoading.value = true;
    memberError.value = null;
    try {
      const response = await invoke<SectorMemberPage>('get_sector_members', {
        kind: target.kind,
        sectorCode: target.code,
        page: memberPage.value,
        pageSize: memberPageSize.value,
        forceRefresh,
      });
      if (request !== memberGeneration) return;
      members.value = response.items;
      memberTotal.value = response.total;
      memberPage.value = response.page;
      memberPageSize.value = response.page_size;
      memberAsOf.value = response.as_of;
      memberSource.value = response.source;
      memberStale.value = response.stale;
    } catch (e) {
      if (request === memberGeneration) memberError.value = `获取成分股失败: ${e}`;
      console.error('[sector] fetchMembers failed:', e);
    } finally {
      if (request === memberGeneration) memberLoading.value = false;
    }
  }

  async function selectMemberPage(next: number) {
    memberPage.value = Math.max(1, next);
    await fetchMembers();
  }

  async function fetchRotation() {
    rotationLoading.value = true;
    rotationError.value = null;
    try {
      const response = await invoke<SectorRotation>('get_sector_rotation');
      const succeeded = new Set(response.statuses.filter(status => status.ok).map(status => status.kind));
      // 失败的类别保留上次成功数据，成功的类别才替换。
      rotationRows.value = [
        ...rotationRows.value.filter(row => !succeeded.has(row.kind)),
        ...response.items,
      ];
      rotationStatuses.value = response.statuses;
      rotationSource.value = `${response.source} · ${response.request_count} 次请求`;
      if (!response.statuses.some(status => status.ok)) {
        rotationError.value = response.statuses.map(status => status.error).filter(Boolean).join('；') || '轮动数据获取失败';
      }
    } catch (e) {
      rotationError.value = `获取板块轮动失败: ${e}`;
      console.error('[sector] fetchRotation failed:', e);
    } finally {
      rotationLoading.value = false;
    }
  }

  return {
    kind,
    keyword,
    page,
    pageSize,
    rows,
    total,
    pageCount,
    asOf,
    source,
    stale,
    loading,
    error,
    selected,
    members,
    memberPage,
    memberPageSize,
    memberTotal,
    memberPageCount,
    memberAsOf,
    memberSource,
    memberStale,
    memberLoading,
    memberError,
    limitUpStats,
    limitUpError,
    rotationRows,
    rotationStatuses,
    rotationSource,
    rotationLoading,
    rotationError,
    fetchSummaries,
    selectKind,
    selectPage,
    selectSector,
    clearSector,
    fetchMembers,
    fetchLimitUpStats,
    selectMemberPage,
    fetchRotation,
  };
});

File: src-tauri/src/commands/sector.rs
use crate::datasource::sector::{
    self, SectorHistory, SectorKind, SectorLimitUpStats, SectorMemberPage, SectorRotation,
    SectorSummaryPage,
};

/// 获取行业或概念板块排行。
///
/// `keyword` 非空时会在板块目录中做全量名称/代码搜索，再按分页返回结果；
/// 普通翻页只请求当前页，避免打开板块中心时拉取所有成分股。
#[tauri::command]
pub async fn get_sector_summaries(
    kind: String,
    page: Option<u32>,
    page_size: Option<u32>,
    keyword: Option<String>,
    force_refresh: Option<bool>,
) -> Result<SectorSummaryPage, String> {
    let kind = SectorKind::parse(&kind)?;
    let keyword: String = keyword
        .unwrap_or_default()
        .trim()
        .chars()
        .take(30)
        .collect();
    sector::fetch_summaries(
        kind,
        page.unwrap_or(1),
        page_size.unwrap_or(50),
        &keyword,
        force_refresh.unwrap_or(false),
    )
    .await
}

/// 按板块代码获取成分股。只接受东方财富板块代码，避免把任意用户输入拼进 URL。
#[tauri::command]
pub async fn get_sector_members(
    kind: String,
    sector_code: String,
    page: Option<u32>,
    page_size: Option<u32>,
    force_refresh: Option<bool>,
) -> Result<SectorMemberPage, String> {
    let kind = SectorKind::parse(&kind)?;
    sector::fetch_members(
        kind,
        &sector_code,
        page.unwrap_or(1),
        page_size.unwrap_or(50),
        force_refresh.unwrap_or(false),
    )
    .await
}

#[tauri::command]
pub async fn get_sector_limit_up_stats(sector_code: String) -> Result<SectorLimitUpStats, String> {
    sector::fetch_limit_up_stats(&sector_code).await
}

/// 按需获取单个板块的日/周/月 K 线，不参与排行和成分股请求。
#[tauri::command]
pub async fn get_sector_history(
    sector_code: String,
    period: String,
) -> Result<SectorHistory, String> {
    sector::fetch_history(&sector_code, &period).await
}

/// 行业、概念各一个批量请求；响应保留两类各自的成功/失败状态。
#[tauri::command]
pub async fn get_sector_rotation() -> Result<SectorRotation, String> {
    Ok(sector::fetch_rotation().await)
}
